src/bentoml/_internal/client/__init__.py

```diff
@@ -30,50 +33,61 @@ def __init__(self, svc: Service, server_url: str):
         self._svc = svc
         self.server_url = server_url
 
-        if svc is not None and len(svc.apis) == 0:
+        if len(svc.apis) == 0:
             raise BentoMLException("No APIs were found when constructing client.")
 
         self.endpoints = []
-        for name, api in self._svc.apis.items():
+        for name in self._svc.apis.keys():
             self.endpoints.append(name)
 
             if not hasattr(self, name):
-                setattr(
-                    self, name, functools.partial(self._sync_call, _bentoml_api=api)
-                )
+                setattr(self, name, functools.partial(self.call, bentoml_api_name=name))
 
             if not hasattr(self, f"async_{name}"):
                 setattr(
                     self,
                     f"async_{name}",
-                    functools.partial(self._call, _bentoml_api=api),
+                    functools.partial(self.async_call, bentoml_api_name=name),
                 )
 
-    def call(self, bentoml_api_name: str, inp: t.Any = None, **kwargs: t.Any) -> t.Any:
-        return self._sync_call(
-            inp, _bentoml_api=self._svc.apis[bentoml_api_name], **kwargs
+    @abstractproperty
+    def sync_client(self) -> SyncClient:
+        raise NotImplementedError()
+
+    @abstractproperty
+    def async_client(self) -> AsyncClient:
+        raise NotImplementedError()
+
+    def call(
+        self,
+        bentoml_api_name: str,
+        inp: t.Any = None,
+        **kwargs: t.Any,
+    ) -> t.Any:
+        return self.sync_client.call(
+            inp=inp, bentoml_api_name=bentoml_api_name, **kwargs
         )
 
     async def async_call(
         self, bentoml_api_name: str, inp: t.Any = None, **kwargs: t.Any
     ) -> t.Any:
-        return await self._call(
-            inp, _bentoml_api=self._svc.apis[bentoml_api_name], **kwargs
+        return await self.async_client.call(
+            inp=inp, bentoml_api_name=bentoml_api_name, **kwargs
         )
 
     @staticmethod
     def wait_until_server_ready(
         host: str, port: int, timeout: float = 30, **kwargs: t.Any
     ) -> None:
         try:
-            from .http import HTTPClient
+            from .http import SyncHTTPClient
 
-            HTTPClient.wait_until_server_ready(host, port, timeout, **kwargs)
+            SyncHTTPClient.wait_until_server_ready(host, port, timeout, **kwargs)
         except BadStatusLine:
             # when address is a RPC
-            from .grpc import GrpcClient
+            from .grpc import SyncGrpcClient
 
-            GrpcClient.wait_until_server_ready(host, port, timeout, **kwargs)
+            SyncGrpcClient.wait_until_server_ready(host, port, timeout, **kwargs)
```

Comment:
nit: For sync client naming would be `GrpcClient`,
and Async would be `AsyncGrpcClient` (similar to what aiohttp does)

One more thing is we might want to fix the sync client `async function` behaviour within a running loop

Reply:
> nit: For sync client naming would be GrpcClient,
> and Async would be AsyncGrpcClient (similar to what aiohttp does)

I chose not to do that since `GrpcClient` already exists and its API is maintained for backwards compatibility.

Reply:
I think most of these should in theory be mostly memory bound and therefore valid usecases for async functions, but we are intending to properly rework IO descriptors (that's pretty much next on my plate, and should hopefully have some preliminary work soon).

Re: making sync functions run in an async environment: I'm not sure that this is a real issue. In an async environment you should really never be calling the sync functions, and it's actively harmful to do so. We should really just create a wrapper around `asyncio.run` that tells users to use the async methods if they're already in a running loop.

Reply:
> Re: making sync functions run in an async environment: I'm not sure that this is a real issue. In an async environment you should really never be calling the sync functions, and it's actively harmful to do so.

I'm not sure this is true in Python, is it? If you write an `async def` function which never calls `await` inside it it will never yield to the event loop, will it? `to_proto` and `from_proto` will never call `await` mostly because Protobuf doesn't have any `async` implementations and doesn't intend to. Making `to/from_proto` `async def` doesn't magically yield to the event loop.

If you don't want to block the event loop I think you'd be way better off making `to/from_proto` synchronous functions and then calling them in `async` functions like this:

```python
executor = concurrent.futures.ThreadPoolExecutor(max_workers=3)
loop = asyncio.get_event_loop()
result = await loop.run_in_executor(executor, self.to_proto, obj)
```

Reply:
Yeah, I'm not entirely sure why the from/to proto functions are async now. Probably just so they matched everything else, I think that's probably just a problem we didn't think about very hard.

Reply:
> I'm not sure this is true in Python, is it? If you write an `async def` function which never calls `await` inside it it will never yield to the event loop, will it?

You're correct, I'm in fact  not sure that a single one of our implementations are actually correctly taking advantage of async, this is somewhat of a mess, and something we need to improve.